flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/reader/TestIcebergSourceSplitReader.java

```diff
@@ -0,0 +1,167 @@
+
+    // Fetch the first batch, and check the result size
+    ArrayBatchRecords<RecordAndPosition<RowData>> result = (ArrayBatchRecords) reader.fetch();
+    assertThat(result.numberOfRecords()).isEqualTo(2);
+
+    // Pause the reading of the split, and check that the read is blocked when trying to fetch new
+    // records
+    reader.pauseOrResumeSplits(ImmutableList.of(split), ImmutableList.of());
+    assertThatThrownBy(
+            () ->
+                Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> reader.fetch() != null))
+        .isInstanceOf(ConditionTimeoutException.class)
+        .hasMessageContaining("was not fulfilled within");
+
+    // Unpause the reading of the split, and check the result size
+    reader.pauseOrResumeSplits(ImmutableList.of(), ImmutableList.of(split));
+    result = (ArrayBatchRecords) reader.fetch();
+    assertThat(result.numberOfRecords()).isEqualTo(3);
+  }
+
+  @Test
+  public void testWakeup() throws Exception {
+    IcebergSourceSplitReader<RowData> reader = reader();
+
+    List<List<Record>> records =
+        ImmutableList.of(
+            RandomGenericData.generate(SCHEMA, 2, 0L), RandomGenericData.generate(SCHEMA, 3, 1L));
+    IcebergSourceSplit split =
+        ReaderUtil.createSplit(records, TEMPORARY_FOLDER, FileFormat.PARQUET, APPENDER_FACTORY);
+
+    // Add the new split to the reader
+    reader.handleSplitsChanges(new SplitsAddition<>(ImmutableList.of(split)));
+
+    // Fetch the first batch, and check the result size
+    ArrayBatchRecords<RecordAndPosition<RowData>> result = (ArrayBatchRecords) reader.fetch();
+    assertThat(result.numberOfRecords()).isEqualTo(2);
+
+    // Pause the reading of the split, and check that the read is blocked when trying to fetch new
+    // records
+    reader.pauseOrResumeSplits(ImmutableList.of(split), ImmutableList.of());
+
+    assertThatThrownBy(
+            () ->
+                Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> reader.fetch() != null))
+        .isInstanceOf(ConditionTimeoutException.class)
+        .hasMessageContaining("was not fulfilled within");
+
+    // Run the new fetch and wakeUp concurrently. Expect empty result from the blocked fetch.
+    ExecutorService executorService =
+        MoreExecutors.getExitingExecutorService(
+            (ThreadPoolExecutor) Executors.newFixedThreadPool(1));
+
+    AtomicBoolean gotEmptyResult = new AtomicBoolean(false);
+    executorService.execute(
+        () -> {
+          try {
+            gotEmptyResult.set(reader.fetch().nextSplit() == null);
+          } catch (IOException e) {
+            throw new RuntimeException(e);
+          }
+        });
+
+    // Wakeup the reading of the split, and check the result size
+    reader.wakeUp();
+    Awaitility.await().atMost(2, TimeUnit.SECONDS).until(() -> gotEmptyResult.get());
+
+    // Unpause and read the data in th next fetch
+    reader.pauseOrResumeSplits(ImmutableList.of(), ImmutableList.of(split));
```

Comment:
it is better to call `wakeUp` here to avoid wait for the 1s timeout. wakeUp is also the expected in the real code.

Reply:
I'm not sure I get this comment.

The `wakeUp` should immediately unblock the `fetch` running by the executor service. The `await` is there to have a timeout if the behaviour is wrong.
I have added the `pollInterval(10, TimeUnit.MILLISECONDS)` on the `await` to make sure that we are polling more frequently than 2 seconds. This could help when some timing issue prevents the `wakeUp` to unblock the fetch immediately - Maybe this is what you were suggesting anyway 😄

Reply:
after this line, we want to unblock the `wakeUp.wait(1000)` to speed up the testing. otherwise, the next `fetch` will wait for 1s.

Reply:
We have a blocking fetch here running in the executor:
```
gotEmptyResult.set(reader.fetch().nextSplit() == null);
```

This one is unblocked by the `wakeUp` here:
```
    // Wakeup the reading of the split, and check the result size
    reader.wakeUp();
```

The `pauseOrResumeSplits` happens after the previous fetch already returned. Here we remove the blocking split, and the next fetch will not trigger a wait.

Do I miss something?

Reply:
after resuming the split from this line, shouldn't `wakeUp` be called?

previous ` Awaitility.await()` was to make sure get return an empty result.

Reply:
Nope. Flink puts the `pauseOrResumeSplits` requests into a task queue, and then issues the `wakeUp`. Once that wakeup makes the `fetch` return, then it feeds the first element of the task queue to the reader. That updates the `pausedSplits`. After that, if there is no more element in the task queue, then `fetch` is called again.

This is the scenario which we replay in this test